lib/open_api_spex/open_api/map.ex

```diff
@@ -0,0 +1,395 @@
+  defp embedded_ref_or_struct(list, mod) when is_list(list) do
+    list
+    |> Enum.map(fn
+      %{"$ref": _} = v -> struct(Reference, v)
+      v -> to_struct(v, mod)
+    end)
+  end
+
+  defp embedded_ref_or_struct(map, mod) when is_map(map) do
+    map
+    |> Map.new(fn
+      {k, %{"$ref": _} = v} ->
+        {k, struct(Reference, v)}
+
+      {k, v} ->
+        {k, to_struct(v, mod)}
+    end)
+  end
+
+  defp convert_value_to_atom_if_present(map, key) do
+    if Map.has_key?(map, key) do
+      Map.update!(map, key, fn
+        nil ->
+          nil
+
+        v ->
+          String.to_atom(v)
+      end)
+    else
+      map
+    end
+  end
+
+  defp prepare_schema(map) do
+    map
+    |> convert_value_to_atom_if_present(:type)
+    |> to_list_of_atoms(:required)
+  end
+
+  defp to_list_of_atoms(map, key) do
+    if Map.has_key?(map, key) do
+      Map.update!(map, key, fn
+        nil ->
+          nil
+
+        v ->
+          v
+          |> Enum.map(&String.to_atom/1)
+      end)
+    else
+      map
+    end
+  end
+
+  defp to_struct(nil, _mod), do: nil
+
+  defp to_struct(tag, Tag) when is_binary(tag), do: tag
+
+  defp to_struct(map, Tag) when is_map(map) do
+    Tag
+    |> struct(map)
+    |> to_struct(:externalDocs, ExternalDocumentation)
+  end
+
+  defp to_struct(list, Tag) when is_list(list) do
+    list
+    |> Enum.map(&to_struct(&1, Tag))
+  end
+
+  defp to_struct(map, Components) do
+    # NOTE: need to conver all keys to strings at a lower level
+    map =
+      Map.new(map, fn
+        {k, v} -> {k, v |> Map.new(fn {k, v} -> {Atom.to_string(k), v} end)}
+      end)
+
+    Components
+    |> struct(map)
+    |> to_struct(:schemas, Schemas)
+    |> to_struct(:responses, Responses)
+    |> to_struct(:parameters, Parameters)
+    |> to_struct(:examples, Examples)
+    |> to_struct(:requestBodies, RequestBodies)
+    |> to_struct(:headers, Headers)
+    |> to_struct(:securitySchemes, SecuritySchemes)
+    |> to_struct(:links, Links)
+    |> to_struct(:callbacks, Callbacks)
+  end
+
+  defp to_struct(map, Link) do
+    Link
+    |> struct(map)
+    |> to_struct(:server, Server)
+    |> to_struct(:requestBody, RequestBody)
+    |> to_struct(:parameters, Parameters)
+  end
+
+  defp to_struct(map, Links), do: embedded_ref_or_struct(map, Link)
+
+  defp to_struct(map, SecurityScheme) do
+    SecurityScheme
+    |> struct(map)
+    |> to_struct(:flows, OAuthFlows)
+  end
+
+  defp to_struct(map, SecuritySchemes), do: embedded_ref_or_struct(map, SecurityScheme)
+
+  defp to_struct(map, OAuthFlow) do
+    struct(OAuthFlow, map)
+  end
+
+  defp to_struct(map, OAuthFlows) do
+    OAuthFlows
+    |> struct(map)
+    |> to_struct(:implicit, OAuthFlow)
+    |> to_struct(:password, OAuthFlow)
+    |> to_struct(:clientCredentials, OAuthFlow)
+    |> to_struct(:authorizationCode, OAuthFlow)
+  end
+
+  defp to_struct(%{"$ref": _} = map, Schema), do: struct(Reference, map)
+
+  defp to_struct(%{type: "number"} = map, Schema) do
```

Comment:
It looks like the four function heads can be combined into one head:

```elixir
  defp to_struct(%{type: type} = map, Schema) when type in ~w(number integer boolean string) do
    map
    |> prepare_schema()
    |> (&struct(Schema, &1)).()
    |> to_struct(:xml, Xml)
  end
```

Reply:
I like that!